File: 백준/Gold/1987. 알파벳/알파벳.cc
#include <bits/stdc++.h>
using namespace std;
int r, c, ret = 0;
char a[24][24];
const int dy[] = {0, 0, 1, -1};
const int dx[] = {1, -1, 0, 0};

int v(int y, int x)
{
    return a[y][x] - 'A';
}
void go(int y, int x, int visited, int cnt)
{
    ret = max(ret, cnt);

    for (int i = 0; i < 4; i++)
    {
        int ny = y + dy[i];
        int nx = x + dx[i];
        if (ny < 0 || nx < 0 || ny >= r || nx >= c)
            continue;
        if (visited & (1 << v(ny, nx)))
            continue;
        go(ny, nx, visited | (1 << v(ny, nx)), cnt + 1);
    }
}
int main()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);

    cin >> r >> c;
    for (int i = 0; i < r; i++)
    {
        for (int j = 0; j < c; j++)
        {
            cin >> a[i][j];
        }
    }

    go(0, 0, 1 << v(0, 0), 1);

    cout << ret << "\n";

    return 0;
}

File: 백준/Gold/1987. 알파벳/알파벳.java
import java.util.Scanner;

public class Main {
    static int r, c, ret;
    static char[][] a;
    static boolean[] visited;
    static final int[] dy = {0, 0, 1, -1};
    static final int[] dx = {1, -1, 0, 0};

    static void go(int y, int x, int sum) {
        ret = Math.max(ret, sum);

        for (int i = 0; i < 4; i++) {
            int ny = y + dy[i];
            int nx = x + dx[i];
            if (ny < 0 || nx < 0 || ny >= r || nx >= c)
                continue;
            int next = a[ny][nx] - 'A';
            if (visited[next])
                continue;
            visited[next] = true;
            go(ny, nx, sum + 1);
            visited[next] = false;
        }
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);

        r = scanner.nextInt();
        c = scanner.nextInt();
        scanner.nextLine(); // Consume newline
        a = new char[r][c];
        visited = new boolean[26]; // Assuming input characters are capital letters from A to Z

        for (int i = 0; i < r; i++) {
            String line = scanner.nextLine();
            for (int j = 0; j < c; j++) {
                a[i][j] = line.charAt(j);
            }
        }

        visited[a[0][0] - 'A'] = true;
        go(0, 0, 1);
        System.out.println(ret);
    }
}
